### PR TITLE
[WIP] Set auth language #654

### PR DESCRIPTION
**Pull request by Ehesp**
This PR gives the ability to set the language for the auth module, or use the device language. Usage is the same as the web sdk:

https://firebase.google.com/docs/auth/web/manage-users#send_a_password_reset_email

- [x] Set language code
- [x] Set language via device default
- [x] PR for documentation
- [x] Get language code
- [x] Updated typescript/flowtype
- [x] Handle dynamic apps
- [x] Tests

### CODE REVIEW

File: lib/modules/auth/index.js [hunk 1]
```diff
@@ -322,6 +322,23 @@ export default class Auth extends ModuleBase {
     return this._native.fetchProvidersForEmail(email);
   }
 
+  /**
+   * Sets the language for the auth module using the device language
+   * @returns {*}
+   */
+  useDeviceLanguage() {
+    return this._native.useDeviceLanguage();
```
Comment (chrisbianca) [inline lib/modules/auth/index.js]: No need to return here or on languageCode

Comment (Ehesp) [reply lib/modules/auth/index.js]: Ah yes!

File: lib/modules/auth/index.js [hunk 2]
```diff
@@ -322,6 +322,23 @@ export default class Auth extends ModuleBase {
     return this._native.fetchProvidersForEmail(email);
   }
 
+  /**
+   * Sets the language for the auth module using the device language
+   * @returns {*}
+   */
+  useDeviceLanguage() {
```
Comment (chrisbianca) [inline lib/modules/auth/index.js]: Can you add some flow typings?

File: lib/index.d.ts [hunk 3]
```diff
@@ -637,6 +637,11 @@ declare module "react-native-firebase" {
          */
         currentUser: User | null
 
+        /**
+         * Gets/Sets the language for the app instance
+         */
+        languageCode: string | null;
```
Comment (Ehesp) [inline lib/index.d.ts]: cc @chrisbianca not sure how to do a getter/setter in TypeScript...? This works for the getter but not sure if it's moan when trying to set it.

Comment (aMarCruz) [reply lib/index.d.ts]: @Ehesp it is ok, `languageCode` is a read/write property, the implementation is hidden.

**Comment by compojoom** [comment]
@ecirish - I said this on discord, but I'll note it here as well. 
It seems that on ios there is an issue - probably not in our code, but it would be good to confirm where the error is.

Using auth.languageCode = 'en' - would sent a firebase email in english. But using auth().useDeviceLanguage() would default to firebase's project default and not the device language.

On Android this seems to behave correctly. It would be good if someone else tests useDeviceLanguage() on iOS as well before we merge this.

**Comment by Ehesp** [comment]
If that is the case, it seems as though it's a Firebase issue? The code on our end is quite simple:

```
[[FIRAuth authWithApp:firApp] useAppLanguage];
```

I'll try and confirm this, but if it is different to Android then we have another discepancy.

**Comment by compojoom** [comment]
I know it seems to be a firebase issue and if it really is - then we can add it to the notes and also submit a bug report to firebase. If we just ship it as it is and it doesn't work, then other people will waste time on this as well :(

**Comment by Ehesp** [comment]
@compojoom We discussed this yesterday internally, for the time being we're just going to support `languageCode`, and add `useDeviceLanguage` as an unsupported method. We've flagged it with Google. 

A workaround would be to use another lib to grab the device language and set it via `languageCode`.

**Comment by compojoom** [comment]
Ok, so when is  3.2.1 coming out :)

File: android/src/main/java/io/invertase/firebase/auth/RNFirebaseAuth.java [hunk 4]
```diff
@@ -1177,6 +1179,31 @@ public void onComplete(@NonNull Task<ProviderQueryResult> task) {
...
+
+    firebaseAuth.setLanguageCode(code);
+  }
+
+  /**
+   * Use the device language
+   * @param appName
+   */
+  @ReactMethod
+  public void useDeviceLanguage(String appName) {
```
Comment (Ehesp) [inline android/src/main/java/io/invertase/firebase/auth/RNFirebaseAuth.java]: @Salakar @chrisbianca should I leave the native ones here for the time being? JS currently flags as unsupported due to Firebase issue, but no harm in them being here?

**Comment by Ehesp** [comment]
@compojoom can confirm the iOS issue! Just need to sort one last thing and should be ready to get merged.

**Comment by Ehesp** [comment]
cc @chrisbianca @Salakar All good to merge is you chaps are happy

File: lib/utils/index.js [hunk 5]
```diff
@@ -340,6 +340,30 @@ export function nativeToJSError(code: string, message: string, additionalProps?:
   return error;
 }
 
+/**
```
Comment (chrisbianca) [inline lib/utils/index.js]: This has moved to lib/utils/native.js so doesn't need to be here anymore - I assume this was from a merge?

File: lib/modules/auth/index.js [hunk 6]
```diff
@@ -342,6 +353,14 @@ export default class Auth extends ModuleBase {
     return this._user;
   }
 
+  get namespace(): string {
```
Comment (chrisbianca) [inline lib/modules/auth/index.js]: Just noticed that this is no longer required either.

**Comment by goleador0322** [comment]
I tried to use useDeviceLanguage function (like firebase.auth().userDeviceLanguage)

but, I got this error.
```
ExceptionsManager.js:65 _reactNativeFirebase2.default.auth(...).useDeviceLanguage is not a function
```

how can i fix this :( ?


Podfile.lock
```
- Firebase/Auth (4.8.1):
    - Firebase/Core
    - FirebaseAuth (= 4.4.2)
  - Firebase/Core (4.8.1):
    - FirebaseAnalytics (= 4.0.7)
    - FirebaseCore (= 4.0.14)
  - FirebaseAnalytics (4.0.7):
    - FirebaseCore (~> 4.0)
    - FirebaseInstanceID (~> 2.0)
    - GoogleToolboxForMac/NSData+zlib (~> 2.1)
    - nanopb (~> 0.3)
  - FirebaseAuth (4.4.2):
    - FirebaseAnalytics (~> 4.0)
    - GoogleToolboxForMac/NSDictionary+URLArguments (~> 2.1)
    - GTMSessionFetcher/Core (~> 1.1)
  - FirebaseCore (4.0.14):
    - GoogleToolboxForMac/NSData+zlib (~> 2.1)
  - FirebaseInstanceID (2.0.8):
    - FirebaseCore (~> 4.0)
```

yarn.lock
```
react-native-firebase@^3.2.2:
  version "3.2.2"
  resolved "https://registry.yarnpkg.com/react-native-firebase/-/react-native-firebase-3.2.2.tgz#768d0384de489ffcee106f53c9fbb41cee7eebb0"
  dependencies:
    bows "^1.6.0"
    opencollective "^1.0.3"
    postinstall-build "^5.0.1"
    prop-types "^15.6.0"
```

**Pull request merged**